dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -260,9 +260,16 @@ BufferedFrame& FramesDecoderGpu::FindEmptySlot() {
   DALI_FAIL("Could not find empty slot in the frame buffer");
 }
 
-
 void FramesDecoderGpu::Reset() {
   SendLastPacket(true);
   FramesDecoder::Reset();
 }
+
+FramesDecoderGpu::~FramesDecoderGpu() {
+  cuvidDestroyVideoParser(nvdecode_state_->parser);
```

Comment:
I think you need to move away `cuvidCreateVideoParser` and `cuvidCreateDecoder` from the `FramesDecoderGpu` constructor. If `cuvidCreateVideoParser` fails, we still have the decoder created but the destructor won't be called.

Reply:
Either that, or use some kind of RAII wrappers for these handles. See example:
```C++
#include <iostream>

using namespace std;

class A {
public:
    A() {
        cout << "A" << endl;
    }
    ~A() {
        cout << "~A" << endl;
    }
};

class Test {
public:
    Test() {
        cout << "Test" << endl;
        throw "wtf";
    }
    ~Test() {
        cout << "~Test" << endl;
    }

    A a;
};

int main() {
    try {
        Test t;
    } catch (...) {}
    return 0;
}
```
If the constructor's body throws, the fields are deemed already constructed, so they will be properly destroyed. The enclosing object, however, is not considered constructed until the constructor successfully returns, so the destructor is not called.

Reply:
Done